Correct nullptr dereference in DocumentLoader::responseReceived

brentfulgham opened the pull request:
#### 97054b3ee7376d125de3be38a569e60db0d7b049
<pre>
Correct nullptr dereference in DocumentLoader::responseReceived
<a href="https://bugs.webkit.org/show_bug.cgi?id=294863">https://bugs.webkit.org/show_bug.cgi?id=294863</a>
&lt;<a href="https://rdar.apple.com/problem/153030488">rdar://problem/153030488</a>&gt;

Reviewed by Charlie Wolfe.

Telemetry from recent betas indicate nullptr crashes in
`DocumentLoader::responseReceived`. This patch corrects this error.

* Source/WebCore/loader/DocumentLoader.cpp:
(WebCore::DocumentLoader::responseReceived): Correct nullptr frame
checks. Also a Drive-by-fix to add missing call to the completion handler.
(WebCore::DocumentLoader::responseReceived): Correct nullptr checks.

Canonical link: <a href="https://commits.webkit.org/296582@main">https://commits.webkit.org/296582@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8aa6908224964912b39ad5916d609d44c88eb4cf

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/108944 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/28604 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/19029 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/114152 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/59285 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/110907 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/29288 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/37171 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/82787 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/111892 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/23285 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/98128 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/63226 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/22704 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/16265 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/58850 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/92659 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/16312 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/117272 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/35994 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/26605 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/91794 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/36366 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/94392 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/91601 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23323 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/36514 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/14270 "Passed tests") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/31860 "Build is in progress. Recent messages:Running apply-patch; Checked out pull request; Running compile-webkit") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/35892 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/41418 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/35593 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/38934 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/37277 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebCore/loader/DocumentLoader.cpp:
```diff
@@ -1000,14 +1001,17 @@ void DocumentLoader::responseReceived(ResourceResponse&& response, CompletionHan
         auto url = response.url();
         RefPtr frame = m_frame.get();
-        ContentSecurityPolicy contentSecurityPolicy(URL { url }, this, frame->document());
+        ReportingClient* reportingClient = nullptr;
```
This seems like a step in the wrong direction.  Why not just use a RefPtr&lt;Document&gt;?

brentfulgham replied:
We don’t want to take ownership of the Document, so perhaps a WeakPtr. After looking at this yesterday, I think this argument to CSP was hacked in a bit poorly, and would need a larger patch to get this right.

I took the approach of matching the logic in the other overload of this method, and intend to follow up with a patch that addresses the ugly bare pointer used for this argument in current code.

cdumez commented on Source/WebCore/loader/DocumentLoader.cpp:
```diff
@@ -1035,12 +1039,15 @@ void DocumentLoader::responseReceived(ResourceResponse&& response, CompletionHan
     m_response = WTFMove(response);
 
     if (m_identifierForLoadWithoutResourceLoader) {
+        RefPtr protectedFrameLoader { frameLoader() };
```
We do not call local variable `protectedXXX` anymore. This is pre-safer CPP style. This should be:
```
RefPtr frameLoader = this->frameLoader();
```

brentfulgham replied:
OK, will fix.

cdumez commented on Source/WebCore/loader/DocumentLoader.cpp:
```diff
@@ -1053,18 +1060,24 @@ void DocumentLoader::responseReceived(ResourceResponse&& response, CompletionHan
         return;
     }
 
+    RefPtr protectedFrame { m_frame.get() };
```
Why do we need this? Why can't we reuse the `frame` local variable in this function?

charliewolfe replied:
It's not in scope?

cdumez replied:
Oh, it's possible. It's hard to read diffs sometimes :)

cdumez replied:
Since we're moving it, we should name it `frame`, not `protectedFrame`.

brentfulgham replied:
Will do!

brentfulgham committed:
Correct nullptr dereference in DocumentLoader::responseReceived
https://bugs.webkit.org/show_bug.cgi?id=294863
<rdar://problem/153030488>

Reviewed by Charlie Wolfe.

Telemetry from recent betas indicate nullptr crashes in
`DocumentLoader::responseReceived`. This patch corrects this error.

* Source/WebCore/loader/DocumentLoader.cpp:
(WebCore::DocumentLoader::responseReceived): Correct nullptr frame
checks. Also a Drive-by-fix to add missing call to the completion handler.
(WebCore::DocumentLoader::responseReceived): Correct nullptr checks.

Canonical link: https://commits.webkit.org/296582@main
